src/com/xilinx/rapidwright/util/CodeGenerator.java

```diff
@@ -0,0 +1,115 @@
+     */
+    public static String testNetGenerator(String dcp, Collection<String> nets, Collection<String> siteInsts) {
+        Design d = Design.readCheckpoint(dcp);
+        return testNetGenerator(d, nets, siteInsts);
+    }
+
+    /**
+     * Generates RapidWright test code from a provided checkpoint that instantiates the provided nets and siteInsts,
+     * and adds the associated PIPs to each of the nets.
+     *
+     * @param dcp The path to the desired DCP to generate code from.
+     * @param edif The path to the external EDIF file to load (in case of an encrypted DCP file).
+     * @param nets A collection of nets to generate code for.
+     * @param siteInsts A collection of site instances to generate code for.
+     * @return The resulting code.
+     */
+    public static String testNetGenerator(String dcp, String edif, Collection<String> nets, Collection<String> siteInsts) {
+        Design d = Design.readCheckpoint(dcp, edif);
+        return testNetGenerator(d, nets, siteInsts);
+    }
+
+    /**
+     * Generates RapidWright test code from a provided checkpoint that instantiates the provided nets and siteInsts,
+     * and adds the associated PIPs to each of the nets.
+     *
+     * @param design The design to generate code from.
+     * @param nets A collection of nets to generate code for.
+     * @param siteInsts A collection of site instances to generate code for.
+     * @return The resulting code.
+     */
+    public static String testNetGenerator(Design design, Collection<String> nets, Collection<String> siteInsts) {
+        StringBuilder code = new StringBuilder();
+        String partName = design.getPartName();
+        String designName = design.getName();
+
+        code.append(String.format("Design design = new Design(\"%s\", \"%s\");\n", designName, partName));
+        code.append("Device device = design.getDevice();\n\n");
+
+        int netIdx = 0;
+        for(String netName : nets) {
+            Net net = design.getNet(netName);
+            String varName = "net";
+            if (nets.size() > 1) varName += netIdx;
+
+            code.append(String.format("Net %s = TestDesignHelper.createTestNet(design, \"%s\", new String[]{\n", varName, varName));
+
+            List<PIP> pips = net.getPIPs();
+            for(int i = 0; i < pips.size(); i++) {
+                PIP pip = pips.get(i);
+                code.append(String.format("        \"%s\"", pip));
+                if (i != pips.size()-1)
+                    code.append(",");
+                code.append("\n");
+            }
+
+            code.append("});\n");
+            netIdx++;
+        }
+        code.append("\n");
+
+        int siteIdx = 0;
+        for(String siteName : siteInsts)
+        {
+            String varName = "si";
+            if (nets.size() > 1) varName += siteIdx;
+
+            code.append(String.format("SiteInst %s = design.createSiteInst(design.getDevice().getSite(\"%s\"));\n", varName, siteName));
+
```

Comment:
```suggestion
            code.append(String.format("SiteInst %s = design.createSiteInst(design.getDevice().getSite(\"%s\"));\n", varName, siteName));
```

Reply:
Although some whitespace is important, let's remove some of the extra newlines throughout this method.